numpy/_core/src/multiarray/iterators.c

```diff
@@ -857,201 +811,154 @@ iter_ass_sub_int(PyArrayIterObject *self, PyArrayObject *ind,
-    npy_intp start, step_size;
-    npy_intp n_steps;
-    PyObject *obj = NULL;
-    NPY_cast_info cast_info = {.func = NULL};
-
     if (val == NULL) {
         PyErr_SetString(PyExc_TypeError,
                 "Cannot delete iterator elements");
         return -1;
     }
 
-    if (PyArray_FailUnlessWriteable(self->ao, "underlying array") < 0)
+    if (PyArray_FailUnlessWriteable(self->ao, "underlying array") < 0) {
         return -1;
+    }
 
-    if (ind == Py_Ellipsis) {
-        ind = PySlice_New(NULL, NULL, NULL);
-        retval = iter_ass_subscript(self, ind, val);
-        Py_DECREF(ind);
-        return retval;
+    int ret = -1;
+
+    int index_type;
+    int index_num = -1;
+    int ndim, fancy_ndim;
+    npy_intp start, stop, step, n_steps;
+    npy_index_info indices[NPY_MAXDIMS * 2 + 1];
+
+    PyArray_Descr *dtype = PyArray_DESCR(self->ao);
+    npy_intp dtype_size = dtype->elsize;
+    NPY_cast_info cast_info = {.func = NULL};
+
+    /* Prepare the indices */
+    index_type = prepare_index_noarray(1, &self->size, ind, indices, &index_num,
+        &ndim, &fancy_ndim, 1, 1);
+
+    if (index_type < 0) {
+        goto finish;
+    }
+
+    else if (indices[0].type == HAS_NEWAXIS) {
+        PyErr_SetString(PyExc_IndexError,
+            "only integers, slices (`:`), ellipsis (`...`) and integer or boolean "
+            "arrays are valid indices"
+        );
+        goto finish;
     }
 
-    if (PyTuple_Check(ind)) {
-        int len;
-        len = PyTuple_GET_SIZE(ind);
-        if (len > 1) {
+    // Single ellipsis index
+    else if (index_type == HAS_ELLIPSIS) {
+        ind = PySlice_New(NULL, NULL, NULL);
+        if (ind == NULL) {
             goto finish;
         }
-        ind = PyTuple_GET_ITEM(ind, 0);
-    }
 
-    type = PyArray_DESCR(self->ao);
+        ret = iter_ass_subscript(self, ind, val);
+        Py_DECREF(ind);
+        goto finish;
+    }
 
-    /*
-     * Check for Boolean -- this is first because
-     * Bool is a subclass of Int
-     */
-   
-    if (PyBool_Check(ind)) {
-        retval = 0;
-        int istrue = PyObject_IsTrue(ind);
-        if (istrue == -1) {
-            return -1;
-        }
-        if (istrue) {
-            retval = PyArray_Pack(
-                    PyArray_DESCR(self->ao), self->dataptr, val);
+    // Single boolean index
+    else if (indices[0].type == HAS_0D_BOOL) {
+        ret = 0;
+        if (indices[0].value) {
+            ret = PyArray_Pack(PyArray_DESCR(self->ao), self->dataptr, val);
         }
         goto finish;
     }
 
-    if (PySequence_Check(ind) || PySlice_Check(ind)) {
-        goto skip;
-    }
-    start = PyArray_PyIntAsIntp(ind);
-    if (error_converting(start)) {
-        PyErr_Clear();
-    }
-    else {
-        if (check_and_adjust_index(&start, self->size, -1, NULL) < 0) {
+    PyArray_ITER_RESET(self);
+
+    if (index_type == HAS_INTEGER) {
+        if (check_and_adjust_index(&indices[0].value, self->size, -1, NULL) < 0) {
```

Comment:
Here we probably need a cast again (or just a test, if there is a test for e.g. `int16` input and it passes, then that would be fine).